sql/core/src/test/scala/org/apache/spark/sql/UnwrapCastInComparisonEndToEndSuite.scala

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.catalyst.expressions.IntegralLiteralTestUtils.{negativeInt, positiveInt}
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.Decimal
+
+class UnwrapCastInComparisonEndToEndSuite extends QueryTest with SharedSparkSession {
+  import testImplicits._
+
+  test("cases when literal is max") {
+    val t = "test_table"
+    withTable(t) {
+      Seq[(Integer, java.lang.Short, java.lang.Float)](
+        (1, 100.toShort, 3.14.toFloat), (2, Short.MaxValue, Float.NaN), (3, null, null))
+        .toDF("c1", "c2", "c3").write.saveAsTable(t)
+      val df = spark.table(t)
+
+      var lit = Short.MaxValue.toInt
+      checkAnswer(df.select("c1").where(s"c2 > $lit"), Seq.empty)
+      checkAnswer(df.select("c1").where(s"c2 >= $lit"), Row(2))
+      checkAnswer(df.select("c1").where(s"c2 == $lit"), Row(2))
+      checkAnswer(df.select("c1").where(s"c2 <=> $lit"), Row(2))
+      checkAnswer(df.select("c1").where(s"c2 != $lit"), Row(1))
+      checkAnswer(df.select("c1").where(s"c2 <= $lit"), Row(1) :: Row(2) :: Nil)
+      checkAnswer(df.select("c1").where(s"c2 < $lit"), Row(1))
+
+      checkAnswer(df.select("c1").where(s"c3 > double('nan')"), Seq.empty)
+      checkAnswer(df.select("c1").where(s"c3 >= double('nan')"), Row(2))
+      checkAnswer(df.select("c1").where(s"c3 == double('nan')"), Row(2))
+      checkAnswer(df.select("c1").where(s"c3 <=> double('nan')"), Row(2))
+      checkAnswer(df.select("c1").where(s"c3 != double('nan')"), Row(1))
+      checkAnswer(df.select("c1").where(s"c3 <= double('nan')"), Row(1) :: Row(2) :: Nil)
+      checkAnswer(df.select("c1").where(s"c3 < double('nan')"), Row(1))
+
+      lit = positiveInt
```

Comment:
this doesn't match the test case name `cases when literal is max`

We can put it in a new test `cases when literal exceeds max`